File: php/features.php
<?php
declare(strict_types=1);

// ImagePlaceholderGenerator SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class ImagePlaceholderGeneratorFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new ImagePlaceholderGeneratorBaseFeature();
            case "ratelimit":
                return new ImagePlaceholderGeneratorRatelimitFeature();
            case "retry":
                return new ImagePlaceholderGeneratorRetryFeature();
            case "test":
                return new ImagePlaceholderGeneratorTestFeature();
            case "timeout":
                return new ImagePlaceholderGeneratorTimeoutFeature();
            default:
                return new ImagePlaceholderGeneratorBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: py/imageplaceholdergenerator_sdk/features.py
# ImagePlaceholderGenerator SDK feature factory

from imageplaceholdergenerator_sdk.feature.base_feature import ImagePlaceholderGeneratorBaseFeature
from imageplaceholdergenerator_sdk.feature.ratelimit_feature import ImagePlaceholderGeneratorRatelimitFeature
from imageplaceholdergenerator_sdk.feature.retry_feature import ImagePlaceholderGeneratorRetryFeature
from imageplaceholdergenerator_sdk.feature.test_feature import ImagePlaceholderGeneratorTestFeature
from imageplaceholdergenerator_sdk.feature.timeout_feature import ImagePlaceholderGeneratorTimeoutFeature


_FEATURES = {
    "base": lambda: ImagePlaceholderGeneratorBaseFeature(),
    "ratelimit": lambda: ImagePlaceholderGeneratorRatelimitFeature(),
    "retry": lambda: ImagePlaceholderGeneratorRetryFeature(),
    "test": lambda: ImagePlaceholderGeneratorTestFeature(),
    "timeout": lambda: ImagePlaceholderGeneratorTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
